File: korus-contracts/scripts/cancel-game.js
const {
  Connection,
  PublicKey,
  Keypair,
  Transaction,
  TransactionInstruction,
  SystemProgram,
  sendAndConfirmTransaction
} = require('@solana/web3.js');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Deployed Program ID on devnet
const PROGRAM_ID = new PublicKey("4iUdAkPRmZLzUFXTLpt5QPGmUUtP6yfgpPpF3sLD9xtd");

async function main() {
  const gameIdArg = process.argv[2];
  const walletPath = process.argv[3];

  if (!gameIdArg || !walletPath) {
    console.error("Usage: node cancel-game.js <GAME_ID> <WALLET_KEYPAIR_PATH>");
    console.error("Example: node cancel-game.js 0 ~/.config/solana/id.json");
    process.exit(1);
  }

  const gameId = parseInt(gameIdArg);

  console.log("🚫 Cancelling game", gameId);
  console.log("=".repeat(50));

  // Load wallet keypair
  const keypairData = JSON.parse(fs.readFileSync(path.resolve(walletPath), 'utf-8'));
  const playerKeypair = Keypair.fromSecretKey(Uint8Array.from(keypairData));

  console.log("Player:", playerKeypair.publicKey.toString());

  // Connect to devnet
  const connection = new Connection("https://api.devnet.solana.com", "confirmed");

  // Get balance
  const balance = await connection.getBalance(playerKeypair.publicKey);
  console.log("Balance:", balance / 1e9, "SOL");

  // Derive PDAs
  const [statePda] = PublicKey.findProgramAddressSync(
    [Buffer.from("state")],
    PROGRAM_ID
  );

  const gameIdBuffer = Buffer.alloc(8);
  gameIdBuffer.writeBigUInt64LE(BigInt(gameId));
  const [gamePda] = PublicKey.findProgramAddressSync(
    [Buffer.from("game"), gameIdBuffer],
    PROGRAM_ID
  );

  const [playerStatePda] = PublicKey.findProgramAddressSync(
    [Buffer.from("player"), playerKeypair.publicKey.toBuffer()],
    PROGRAM_ID
  );

  const [escrowPda] = PublicKey.findProgramAddressSync(
    [Buffer.from("escrow"), gamePda.toBuffer()],
    PROGRAM_ID
  );

  console.log("\nPDAs:");
  console.log("  State:", statePda.toString());
  console.log("  Game:", gamePda.toString());
  console.log("  Player State:", playerStatePda.toString());
  console.log("  Escrow:", escrowPda.toString());

  // Create cancel_game instruction
  // Discriminator: SHA256("global:cancel_game")[:8]
  const discriminatorHash = crypto.createHash('sha256').update('global:cancel_game').digest();
  const discriminator = discriminatorHash.slice(0, 8);

  console.log("\nDiscriminator:", Array.from(discriminator));

  const instruction = new TransactionInstruction({
    programId: PROGRAM_ID,
    keys: [
      { pubkey: statePda, isSigner: false, isWritable: true },
      { pubkey: gamePda, isSigner: false, isWritable: true },
      { pubkey: playerStatePda, isSigner: false, isWritable: true },
      { pubkey: escrowPda, isSigner: false, isWritable: true },
      { pubkey: playerKeypair.publicKey, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    data: discriminator,
  });

  // Create and send transaction
  const transaction = new Transaction().add(instruction);

  try {
    console.log("\n📤 Sending cancel transaction...");
    const signature = await sendAndConfirmTransaction(
      connection,
      transaction,
      [playerKeypair],
      { commitment: 'confirmed' }
    );

    console.log("✅ Game cancelled successfully!");
    console.log("Transaction signature:", signature);
    console.log("Explorer:", `https://explorer.solana.com/tx/${signature}?cluster=devnet`);
  } catch (error) {
    console.error("\n❌ Failed to cancel game:", error.message);
    if (error.logs) {
      console.log("\n📋 Program logs:");
      error.logs.forEach(log => console.log("  ", log));
    }
  }
}

main().catch(console.error);

File: korus-contracts/scripts/cancel-game.py
#!/usr/bin/env python3
"""
Cancel a game on the Solana blockchain
Usage: python3 cancel-game.py <game_id> <wallet_keypair_path>
"""

import sys
import json
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.transaction import Transaction
from solders.instruction import Instruction, AccountMeta
from solders.system_program import ID as SYS_PROGRAM_ID
from solana.rpc.api import Client
import hashlib

PROGRAM_ID = Pubkey.from_string("4iUdAkPRmZLzUFXTLpt5QPGmUUtP6yfgpPpF3sLD9xtd")

def get_discriminator(name: str) -> bytes:
    """Calculate Anchor instruction discriminator"""
    return hashlib.sha256(name.encode()).digest()[:8]

def find_program_address(seeds: list, program_id: Pubkey) -> tuple:
    """Find PDA for given seeds"""
    seed_bytes = []
    for seed in seeds:
        if isinstance(seed, str):
            seed_bytes.append(seed.encode())
        elif isinstance(seed, bytes):
            seed_bytes.append(seed)
        elif isinstance(seed, int):
            seed_bytes.append(seed.to_bytes(8, 'little'))

    return Pubkey.find_program_address(seed_bytes, program_id)

def main():
    if len(sys.argv) != 3:
        print("Usage: python3 cancel-game.py <game_id> <wallet_keypair_path>")
        sys.exit(1)

    game_id = int(sys.argv[1])
    keypair_path = sys.argv[2]

    print(f"🚫 Cancelling game {game_id}")
    print("=" * 50)

    # Load wallet keypair
    with open(keypair_path, 'r') as f:
        keypair_data = json.load(f)
    player_keypair = Keypair.from_bytes(bytes(keypair_data))
    player_pubkey = player_keypair.pubkey()

    print(f"Player: {player_pubkey}")

    # Connect to devnet
    client = Client("https://api.devnet.solana.com")

    # Get balance
    balance_resp = client.get_balance(player_pubkey)
    balance = balance_resp.value / 1e9
    print(f"Balance: {balance} SOL")

    # Derive PDAs
    state_pda, _ = find_program_address([b"state"], PROGRAM_ID)
    game_pda, _ = find_program_address([b"game", game_id], PROGRAM_ID)
    player_state_pda, _ = find_program_address([b"player", bytes(player_pubkey)], PROGRAM_ID)
    escrow_pda, _ = find_program_address([b"escrow", bytes(game_pda)], PROGRAM_ID)

    print("\nPDAs:")
    print(f"  State: {state_pda}")
    print(f"  Game: {game_pda}")
    print(f"  Player State: {player_state_pda}")
    print(f"  Escrow: {escrow_pda}")

    # Create cancel_game instruction
    discriminator = get_discriminator("global:cancel_game")
    print(f"\nDiscriminator: {list(discriminator)}")

    instruction = Instruction(
        program_id=PROGRAM_ID,
        accounts=[
            AccountMeta(pubkey=state_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=game_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=player_state_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=escrow_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=player_pubkey, is_signer=True, is_writable=True),
            AccountMeta(pubkey=SYS_PROGRAM_ID, is_signer=False, is_writable=False),
        ],
        data=discriminator,
    )

    # Create and send transaction
    recent_blockhash = client.get_latest_blockhash().value.blockhash
    transaction = Transaction.new_with_payer([instruction], player_pubkey)
    transaction.recent_blockhash = recent_blockhash
    transaction.sign([player_keypair])

    try:
        print("\n📤 Sending cancel transaction...")
        result = client.send_transaction(transaction)
        signature = result.value

        print("✅ Game cancelled successfully!")
        print(f"Transaction signature: {signature}")
        print(f"Explorer: https://explorer.solana.com/tx/{signature}?cluster=devnet")
    except Exception as e:
        print(f"\n❌ Failed to cancel game: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
